aea/package_manager/base.py

```diff
@@ -0,0 +1,426 @@
+        :param update_hashes: Update hashes in the packages.json if the calculated
+                              hash for a package does not match the one in the
+                              packages.json.
+        """
+
+        sync_needed = False
+        hash_updates = packages.copy()
+        package_updates = OrderedDict()
+
+        for package_id in packages:
+            package_path = self.package_path_from_package_id(package_id)
+
+            if package_path.exists():
+                package_hash = IPFSHashOnly.get(str(package_path))
+                expected_hash = packages[package_id]
+
+                if package_hash == expected_hash:
+                    continue
+
+                if update_hashes:
+                    hash_updates[package_id] = package_hash
+                    continue
+
+                if update_packages:
+                    sync_needed = True
+                    package_updates[package_id.with_hash(expected_hash)] = expected_hash
+                    self._logger.info(
+                        f"Hash does not match for {package_id}, downloading package again..."
+                    )
+                    continue
+
+                raise PackageHashDoesNotMatch(
+                    f"Hashes for {package_id} does not match; Calculated hash: {package_hash}; Expected hash: {expected_hash}"
+                )
+
+            sync_needed = True
+            self._logger.info(f"{package_id} not found locally, downloading...")
+            package_id_with_hash = package_id.with_hash(packages[package_id])
+            self.add_package(package_id=package_id_with_hash)
+
+        return sync_needed, hash_updates, package_updates
+
+    def sync(
+        self,
+        dev: bool = False,
+        third_party: bool = True,
+        update_packages: bool = False,
+        update_hashes: bool = False,
+    ) -> "PackageManager":
+        """Sync local packages to the remote registry."""
+
+        if update_packages and update_hashes:
+            raise ValueError(
+                "Both `update_packages` and `update_hashes` cannot be set to `True`."
+            )
+
+        self._logger.info(f"Performing sync @ {self.path}")
+
+        sync_needed = False
+
+        if third_party:
+            self._logger.info("Checking third party packages.")
+            (
+                _sync_needed,
+                hash_updates_third_party,
+                package_updates_third_party,
+            ) = self._sync(
+                packages=self.third_party_packages,
+                update_hashes=update_hashes,
+                update_packages=update_packages,
+            )
+            sync_needed = sync_needed or _sync_needed
+
+            if update_hashes and hash_updates_third_party:
+                self._logger.warning(
+                    f"Hashes for follwing third party module has changed.\n\t- "
+                    + "\n\t- ".join(map(str, hash_updates_third_party))
+                )
```

Comment:
If hash change is detected for a third party we won't update the packages unless the user wants to do it explicitly

Reply:
this thus resolves issue arising from me altering a third party package and messing up my other hashes